packages/aws-library/src/aws_library/s3/_client.py

```diff
@@ -0,0 +1,460 @@
+        await self._exit_stack.aclose()
+
+    async def http_check_bucket_connected(self, *, bucket: S3BucketName) -> bool:
+        with log_catch(_logger, reraise=False):
+            return await self.bucket_exists(bucket=bucket)
+        return False
+
+    @s3_exception_handler(_logger)
+    async def create_bucket(
+        self, *, bucket: S3BucketName, region: BucketLocationConstraintType
+    ) -> None:
+        with log_context(
+            _logger, logging.INFO, msg=f"Create bucket {bucket} in {region}"
+        ):
+            try:
+                # NOTE: see https://github.com/boto/boto3/issues/125 why this is so... (sic)
+                # setting it for the us-east-1 creates issue when creating buckets
+                create_bucket_config: dict[str, Any] = {"Bucket": f"{bucket}"}
+                if region != _DEFAULT_AWS_REGION:
+                    create_bucket_config["CreateBucketConfiguration"] = {
+                        "LocationConstraint": region
+                    }
+
+                await self._client.create_bucket(**create_bucket_config)
+
+            except self._client.exceptions.BucketAlreadyOwnedByYou:
+                _logger.info(
+                    "Bucket %s already exists and is owned by us",
+                    bucket,
+                )
+
+    @s3_exception_handler(_logger)
+    async def bucket_exists(self, *, bucket: S3BucketName) -> bool:
+        """
+        :raises: S3AccessError for any other error
+        """
+        try:
+            await self._client.head_bucket(Bucket=bucket)
+            return True
+        except botocore_exc.ClientError as exc:
+            status_code = exc.response.get("Error", {}).get("Code", -1)
+            if status_code == "404":
+                return False
+            raise
+
+    @s3_exception_handler(_logger)
+    async def object_exists(
+        self, *, bucket: S3BucketName, object_key: S3ObjectKey
+    ) -> bool:
+        # SEE https://www.peterbe.com/plog/fastest-way-to-find-out-if-a-file-exists-in-s3
+        response = await self._client.list_objects_v2(Bucket=bucket, Prefix=object_key)
+        return len(response.get("Contents", [])) > 0
+
+    @s3_exception_handler(_logger)
+    async def get_object_metadata(
+        self, *, bucket: S3BucketName, object_key: S3ObjectKey
+    ) -> S3MetaData:
+        response = await self._client.head_object(
+            Bucket=bucket, Key=object_key, ChecksumMode="ENABLED"
+        )
+        return S3MetaData.from_botocore_head_object(object_key, response)
+
+    @s3_exception_handler(_logger)
+    async def get_directory_metadata(
+        self, *, bucket: S3BucketName, prefix: str
+    ) -> S3DirectoryMetaData:
+        size = 0
+        async for s3_object in self._list_all_objects(bucket=bucket, prefix=prefix):
+            size += s3_object.size
+        return S3DirectoryMetaData(size=size)
+
+    @s3_exception_handler_async_gen(_logger)
+    async def list_objects_paginated(
```

Comment:
MINOR: On the one hand, typically, functions that return async generators are prefixed with `aiter_`    (e.g. `aiter_rows`, `aiter_cols`, etc).

On the other hand, `pagination` suggests a returned **object** that includes metadata, links and also the items. while here every iteration will return a `list[S3MetaData]`. 

My suggestions is to name it `async def iter_objects_chunk` or `async def iter_objects_list` or `async def iter_list_of_objects`

Reply:
generally speaking I kind of dislike repeating in the names what the typing already tells us.
when I hover on the method I see it's a generator, why name it with iter again?

Reply:
already paginated I find it a bit unnecessary to be honest, since it's a generator